### ISSUE TITLE
feat(model): native EAGLE-3 drafter with penultimate feature extrapolation and tree proposal generation

### ISSUE THREAD

**Issue by anthony-chaudhary** [anthony-chaudhary/fak#12728]
# feat(model): native EAGLE-3 drafter with penultimate feature extrapolation and tree proposal generation

```routing
lane: internal/model
paths: internal/model/eagle3.go, internal/model/eagle3_test.go
expected_steps: 5
```

## Lane

`internal/model` — inference runtime, proposal generators, and speculative decoding algorithms.

## Problem frame

- Centrality: Core
- Portfolio tier: 2 (serving)
- Priority: P1
- P1: advanced - extrapolates target model penultimate hidden states to propose high-accuracy candidate trees
- P2: preserved - guarantees zero output token drift and exact distribution preservation under verification
- P3: advanced - constructs candidate trees with dynamic branching topologies and draft-to-target vocabulary remapping
- P4: preserved - feeds tree proposals directly into target batch verification without format conversion

## Parent context

Part of the Native Speculative Acceleration Series (`docs/tickets/native-speculative-acceleration/INDEX.md`), delivering the state-of-the-art EAGLE-3 feature-level tree drafting architecture.

## Current state

Existing speculative drafting in `internal/model/speculative.go` relies on either Multi-Token Prediction heads (`MTPProposalGenerator`), compact sidecar models (`DraftModelProposalGenerator`), or token n-grams (`NGramProposalGenerator`). All three methods operate purely on discrete token embeddings.

EAGLE-3 (Extrapolation Algorithm for Greater Language-model Efficiency v3) achieves significantly higher acceptance rates (mean acceptance length > 3.0 tokens) by operating at the **feature representation level**. Instead of predicting tokens directly from token embeddings, EAGLE-3 captures the hidden states from the penultimate layer of the target model, fuses them with current token embeddings, and passes them through a single lightweight transformer decoder layer. Furthermore, EAGLE-3 generates non-linear candidate trees (e.g. 16 to 32 nodes across 4 to 6 depth steps) rather than single linear token chains. The `fak` runtime currently lacks a native EAGLE-3 proposal engine and feature-fusion pipeline.

## Why this is next

EAGLE-3 represents the state of the art in speculative decoding acceptance rates across diverse domains (reasoning, coding, structured JSON, creative writing). By combining penultimate feature extrapolation with multi-branch candidate trees, EAGLE-3 delivers 2.5x to 3.5x speedups on modern LLMs (LLaMA-3, Qwen-2.5/3, GLM) while preserving bit-level target model equivalence.

## Working spine

Target model prefill or decode step emits penultimate hidden state $H_L$ -> `Eagle3ProposalGenerator` fuses $H_L$ with token embedding -> single-layer transformer head predicts next feature vector $H_{L+1}$ and candidate token logits -> tree expansion module iteratively expands top-k branches -> constructs `DraftProposal` with `CandidateTree` topology, non-sequential position IDs, and 2D tree attention mask -> dispatches to `TargetVerificationEngine`.

## Core through-line

1. Implement `Eagle3ProposalGenerator` conforming to `ProposalGenerator` in `internal/model/eagle3.go`.
2. Implement the EAGLE-3 feature fusion layer that projects the target model's penultimate layer hidden state and concatenates/adds it to the draft token embedding vector.
3. Author the single-layer EAGLE-3 transformer forward pass: self-attention over draft features, LayerNorm/RMSNorm, and feed-forward projection yielding next-step hidden features and vocabulary logits.
4. Implement draft-to-target vocabulary remapping: if the EAGLE-3 head uses a compact or quantized vocabulary, map logits onto the target model token vocabulary using an indexed scatter-gather projection.
5. Author the tree expansion builder: recursively expand candidate nodes up to a configurable budget (e.g. 16 or 32 nodes, depth 4–6), generating the complete `CandidateTree` structure, parent pointers, depth position IDs, and 2D tree attention mask.
6. Provide unit and regression tests asserting valid tree DAG construction (acyclic, valid parent references), feature fusion numerical stability, and strict output sequence agreement with greedy target generation.

## Gold-plating boundary

Strictly excluded from this work unit:
- Training EAGLE-3 feature extrapolation weights from scratch (loads pre-trained weights).
- Multi-layer recurrent draft decoders exceeding the standard single-layer EAGLE-3 architecture.
- Heterogeneous multi-modal draft feature projections (confined to text LLM hidden states).

## Verifiable Witness

```bash
go test -v ./internal/model -run 'TestEagle3ProposalGenerator' -count=1
```

## Witness

```bash
go test -v ./internal/model -run 'TestEagle3ProposalGenerator' -count=1
```

### File:Line Seam

- `internal/model/speculative.go:20 (ProposalGenerator)`
- `internal/model/eagle3.go:1 (Eagle3ProposalGenerator)`
- `internal/model/speculative_verify.go:1 (TargetVerificationEngine)`
- `internal/model/transformer.go:50 (Transformer)`

### Blast Radius & Affected Lanes

- **Primary Lane:** `internal/model`.
- **Unaffected Lanes:** `internal/compute`, `internal/agent`, `platform/*` remain unaffected.

### Quarantined Fallback Mechanism

If EAGLE-3 weights are missing, feature dimensions mismatch the base model, or tree generation encounters an error, the engine seamlessly falls back to MTP or prompt n-gram proposals with zero client-visible interruption `[SW-VERIFIED]`.

## Scoped Acceptance Criteria

- [ ] 1. [SW-VERIFIED] Implement `Eagle3ProposalGenerator` in `internal/model/eagle3.go` supporting penultimate feature fusion and single-layer transformer feature extrapolation.
- [ ] 2. [SW-VERIFIED] Implement candidate tree expansion constructing a valid DAG with exact parent references, position IDs, and 2D tree attention mask.
- [ ] 3. [SW-VERIFIED] Implement draft-to-target vocabulary mapping when draft vocabulary dimension differs from target vocabulary.
- [ ] 4. [SW-VERIFIED] Target model verification tests verify exact sequence identity when verifying EAGLE-3 candidate trees against standard greedy generation.
- [ ] 5. [HW-WITNESSED] Verify on physical hardware that EAGLE-3 tree drafting achieves >= 2.5x decode throughput acceleration over standard autoregressive baseline.

## Done condition

- [ ] 1. [SW-VERIFIED] Implement `Eagle3ProposalGenerator` in `internal/model/eagle3.go` supporting penultimate feature fusion and single-layer transformer feature extrapolation.
- [ ] 2. [SW-VERIFIED] Implement candidate tree expansion constructing a valid DAG with exact parent references, position IDs, and 2D tree attention mask.
- [ ] 3. [SW-VERIFIED] Implement draft-to-target vocabulary mapping when draft vocabulary dimension differs from target vocabulary.
- [ ] 4. [SW-VERIFIED] Target model verification tests verify exact sequence identity when verifying EAGLE-3 candidate trees against standard greedy generation.
- [ ] 5. [HW-WITNESSED] Verify on physical hardware that EAGLE-3 tree drafting achieves >= 2.5x decode throughput acceleration over standard autoregressive baseline.

## Acceptance gate

`go test -v ./internal/model -run 'TestEagle3ProposalGenerator' -count=1` exits 0 with 0 boundary violations.

## Closure binding

Resolving commit carries `(fak model)` trailer.

## Likely files

- `internal/model/eagle3.go`
- `internal/model/eagle3_test.go`
- `internal/model/speculative.go`
- `internal/model/speculative_verify.go`

## Expected steps

5
